Support healthcheck

marusak opened the pull request:
Fixes #345
One known limitation is reported in https://github.com/containers/podman/issues/13083

TODO - if image has healthcheck defined, prefill the values

---

# Support for health checks

It is now possible to set up, inspect and trigger health checks for container. Also the most recent history as well as overall status are presented.

![Screenshot from 2022-02-04 10-41-59](https://user-images.githubusercontent.com/12330670/152507040-c46a8354-9c2d-43be-977a-f603c8eca0db.png)

marusak commented:
@garrett WDYT of the design?

jelly commented:
Do we really need to have the health check details this prominent in the details view? Can't they be part of the `Health check` logs?

marusak commented:
> Do we really need to have the health check details this prominent in the details view? Can't they be part of the Health check logs?

Something for design review. But I think there is enough space for it so I would not try to hide it.

marusak commented on src/ImageRunModal.jsx:
```diff
@@ -1034,6 +1052,99 @@ export class ImageRunModal extends React.Component {
+                                        step={1}
+                                        min={0}
+                                        max={262144}
+                                        onChange={value => this.onValueChanged('healthcheck_interval', parseInt(value))} />
```
0 means disabled

marusak commented:
@jelly I updated your comments. I also changed the layout - now everything is under new `Health check` tab. I updated screenshot in the first comment.
Commits are mess, they need squashing and also there are some unrelated changes. I'll cherry-pick those. Needs also rebase, but for that I want to first stabilize design.

jelly commented on src/ContainerHealthLogs.jsx:
```diff
@@ -0,0 +1,117 @@
+            </Flex>
+            <ListingTable aria-label={_("Logs")}
+                          variant='compact'
+                          columns={['Last 5 runs', _("Started at")]}
```
`Last 5 Runs` should be translated as well right?

marusak replied:
Indeed!

jelly commented on webpack.config.js:
```diff
@@ -25,14 +28,14 @@ const copy_files = [
+if (eslint) {
+    plugins.push(new ESLintPlugin({ extensions: ["js", "jsx"], failOnWarning: true, }));
+}
+
```
hmm why are these changes here? Ideally this should go through starterkit first right?

marusak replied:
well, it is already in cockpit anyway :) We can add it into starter kit, but it seems like pattern that you don't want people to use, it is just good hack when you quickly iterate over the code

garrett commented:
Oh, apologies for not getting back to this. I need to help out with the design a bit. I guess next week?

marusak commented:
> When there's no output, we probably should probably not have expanders on the rows:

Can take a look. Just mind that right now due to https://github.com/containers/podman/issues/13083 no runs have any output, but it should be very soon available. (I am not sure if we can mix and match these rows with and without expanders)

> What happens after 5 runs? We just show the latest ones? Would it be useful to know when something went wrong?

That is limitation from podman itself. It only stores and exposes last 5 runs. If you have 5 runs shown and a new one happens, the oldest one disappears.

> The order is going from first to last, which is quite wrong if we're truncating the information. It should be latest first, historical after. (Flip the order.)

Will do.

> Should we consider something other than a list, especially if it's just a bunch of successes? I guess if there's detailed information, it might be interesting... but a health check is probably only interesting if it passes or when it goes wrong, correct?

Right, if it is bunch of successes you probably don't care that much - maybe just to check if it actually does something or it dies silently. But at the same time, when you see your container is `Healthy` in overview, you likely won't even visit the page.

marusak commented:
@garrett I've updaed your comments. It does not have toggle if there is not output, I switched the order and only show buttons to run healtcheck when the container is running.

![Screenshot from 2022-02-23 08-26-43](https://user-images.githubusercontent.com/12330670/155276791-102f0749-2401-42af-ab40-a1bb54feb4d3.png)

garrett commented:
The defaults for health check are every 30 seconds. Is this a default from somewhere? Is this what we want? The logs have duplicate times as a result:

![image](https://user-images.githubusercontent.com/10246/155292402-ea72b217-edf7-41f8-b095-4128a622a0ed.png)

The labels should be in a label group:

https://www.patternfly.org/v4/components/label-group/

![image](https://user-images.githubusercontent.com/10246/155292715-2ce9627b-f7e1-4bf6-b548-705364f9a643.png)

We also still have the overflow issue, which is even more noticeable with the badge-y labels like this.

marusak commented:
> The defaults for health check are every 30 seconds. Is this a default from somewhere? 

Yes, from `man podman run`:
```
   --health-interval=interval
       Set an interval for the healthchecks. An interval of disable results in no automatic timer setup. The default is 30s.
```

> Is this what we want?

Yes, we don't want to create some arbitrary defaults on our own.

> The logs have duplicate times as a result:

Let me take a look if showing seconds makes sense.

marusak commented:
> > The logs have duplicate times as a result:

> Let me take a look if showing seconds makes sense.

Hmm, we use https://date-fns.org/v2.28.0/docs/formatRelative which does not show seconds. We would have to switch to some different method and would loose the relative part.

marusak commented:
> The labels should be in a label group:

Updated

garrett reviewed (changes requested):
Firstly, I was playing around with the idea of not stretching the items and making the column count 4 (which can't be done in PF as a modifier, but can be adjusted via a variable).

```scss
.container-details-healthcheck {
    --pf-c-description-list--GridTemplateColumns--count: 4;
}
```

There are a few odd things here:

1. "Starting health check" wraps as it's a long label. We should consider something smaller or do something else.
2. Health check status isn't shown outside of the icon. We should have success / fail text.

![Screenshot 2022-03-01 at 14-02-01 Podman containers - garrett Bolt](https://user-images.githubusercontent.com/10246/156173667-625c8675-13b6-4375-aaa7-05e9a360f342.png)

(The names were generated; the naughty one happens to be naughty.)

If I bump the column count to 5, the items don't wrap. This even lets the items line up across expanded containers:

![Screenshot 2022-03-01 at 14-06-34 Podman containers - garrett Bolt](https://user-images.githubusercontent.com/10246/156174436-18c2613f-3165-4a18-b4e5-13ef53641a7a.png)

So we should definitely fix the "starting" label and the health check status. And optionally consider not stretching the healthcheck columns.

garrett commented:
Also, I noticed that the alignment of the flex that wraps the health check info and the button isn't correct. The items should be aligned to the start with an `align-items: start` or `align-items: stretch`, the latter of which I think can be set in PF?

garrett commented:
Does this pre-fill the suggested healthcheck from the container?

`Dockerfile` / `Containerfile` supports a `HEALTHCHECK` directive; it's not just at container creation time with command-line flags:

https://docs.docker.com/engine/reference/builder/#healthcheck

marusak commented:
I tried to prepare such image for myslef, but it says:
```
WARN[0000] Healthcheck is not supported for OCI image format and will be ignored. Must use `docker` format.
```

Adding `--format docker` option to `podman build` now contains it, so I guess I could show it if it is defined, but I think it is going to be a very rear occasion.

marusak commented:
> Adding --format docker option to podman build now contains it, so I guess I could show it if it is defined, but I think it is going to be a very rear occasion.

Also API does not expose it :man_shrugging:

marusak commented:
> "Starting health check" wraps as it's a long label. We should consider something smaller or do something else.

Right, it depends on screen size. It is never gonna be ideal as different translation etc still might be too long. We maybe could prevent manually any kind of wrapping?

marusak commented:
> Health check status isn't shown outside of the icon. We should have success / fail text.

Added

garrett commented:
Hi! What's the status of this PR?

---

I built it and I just get an "**Ooops!**" when trying it out.

Error:

```js
TypeError: containerDetail.State.Healthcheck is undefined
    renderRow Containers.jsx:389
    rows Containers.jsx:653
    card Containers.jsx:652
    render Containers.jsx:650
    React 12
    unstable_runWithPriority scheduler.development.js:468
    React 6
    updateState app.jsx:118
    inspectContainerDetail app.jsx:147
[app.jsx:149:36](webpack://podman/src/app.jsx)
```

This is after doing a git clean, making sure the branch was updated, etc.

marusak commented:
Oh my, the API changed and in fedora-36/rhel-8-6 it is `State.Health` but in all previous ones it is `State.Healthcheck`. $#@! stable podman API! :laughing: 

https://github.com/containers/podman/commit/1199733754b1185d0d2a7a3d832c57565864d0f4

jelly commented on src/client.js:
```diff
@@ -114,6 +114,14 @@ export function commitContainer(system, commitData) {
 
+export function runHealthcheck(system, id) {
+    return new Promise((resolve, reject) => {
+        podmanCall("libpod/containers/" + id + "/healthcheck", "GET", {}, system)
```
This is already a promise so you can return it directly.

marusak replied:
Updated

marusak commented:
ubuntu-stable does not get event as other OSes when new health check is run. Looking if there is some way how we can recognize this.

marusak commented on src/ImageRunModal.scss:
```diff
@@ -77,3 +77,11 @@
     }
 }
+
+.healthcheck-seconds {
+    max-width: 25ch;
+}
+
+.healthcheck-retries {
+    max-width: 9ch;
+}
```
I hate this. I will use `<NumberInput>` where the width can be controlled.

marusak replied:
changed

marusak commented:
@garrett I could not find any PF options to make the list wrap and not fill the full width. I put down a small hack and now it behaves as we expected it to. Is this ok?

![Screenshot from 2022-05-16 12-35-17](https://user-images.githubusercontent.com/12330670/168574881-155608cd-67b2-400d-ad6a-2803e44cabd3.png)

Also now the modal looks like this:
![Screenshot from 2022-05-16 12-35-32](https://user-images.githubusercontent.com/12330670/168574926-221c8d9e-df09-4fa0-842b-54d85bab4965.png)

garrett commented:
I'm hitting a few errors when testing out this PR. I think some are not from this PR, but I do see that when a container is running and has a health check. I'll try to see what's causing the issues here in a bit. If I can reproduce, I'll open an issue or post here (if it's relevant to this PR).

When you stop the container, the health check tab goes away. I guess it's only available when containers are running? Is there any way to know? Ideally, we'd show the past health checks, but even if health checks exist and we can't read them, then we could keep the tab around and say something about how health checks are not available unless the container is running.

garrett commented:
![Screenshot 2022-05-16 at 14-49-07 Podman containers - garrett@Bolt](https://user-images.githubusercontent.com/10246/168596163-79a8673c-b0ac-4f6e-a5da-deddf7d9cd09.png)

Tabs and content, including the current and list, are too wide and flowing off the side. We really need the tab overflow that I've requested from PF for quite some time.

There's an override somewhere in Cockpit — not sure if it's in the normal overrides or a local override — for tabs in the expanded tables. I think it's in the normal overrides, which we're not using in Cockpit-Podman at the moment (but should be), as seen by the gutters in mobile mode (which shouldn't be there).

Fixing the tabs _might_ fix the content. It may be flowing to be too wide as a result of the tabs making the content too wide. Or it could be another issue.

The tabs issue is also a problem for the create dialog in mobile sizes:

![Screenshot 2022-05-16 at 14-53-12 Podman containers - garrett@Bolt](https://user-images.githubusercontent.com/10246/168597097-c43f4268-759c-47e6-b0a1-b40895b443d6.png)

![Screenshot 2022-05-16 at 14-53-32 Podman containers - garrett@Bolt](https://user-images.githubusercontent.com/10246/168597100-39da0942-a8a7-4a4f-aded-eca0cebc7a33.png)

marusak commented:
> When you stop the container, the health check tab goes away. I guess it's only available when containers are running? Is there any way to know? Ideally, we'd show the past health checks, but even if health checks exist and we can't read them, then we could keep the tab around and say something about how health checks are not available unless the container is running.

![Screenshot from 2022-05-17 13-16-24](https://user-images.githubusercontent.com/12330670/168798855-60c24f8e-ef2f-45ec-a203-138ab4bc22e3.png)


seems to work here just fine :thinking: Can you reproduce this?

garrett commented:
> seems to work here just fine thinking Can you reproduce this?

Yes, it happens for every container I've tried here. I can try again now.

garrett commented:
All the issues I could quickly find related to this specific tab overflow problem I've brought up with PF over the past several years (and there might be more than just this):

- https://github.com/patternfly/patternfly/issues/1625
- https://github.com/patternfly/patternfly/pull/2757
- https://github.com/patternfly/patternfly/issues/4800
- https://github.com/patternfly/patternfly-design/issues/840
- https://github.com/patternfly/patternfly-design/issues/1034
- https://github.com/cockpit-project/cockpit-podman/issues/845

And _none_ of these have a solution for this particular problem yet. (Some solve some other tab related problems in different ways for other teams.)

We side-stepped the problem for Cockpit-Machines by _completely redesigning and overhauling the page_ as to not use tabs in expandable lists, but sub-pages instead. (Yes, this was one of the reasons, but it wasn't the only one. There were other good reasons as to why, mainly related to making a machine visible and editable all in one go instead of forcing the user to hunt around for information and for us to be constrained layout-wise.) And in other places in Cockpit, we've managed to keep the number low, like with two tabs in most other place. So it became less of a priority. But here it is again.

[Our workaround is for tertiary navigation](https://github.com/cockpit-project/cockpit/blob/main/pkg/lib/patternfly/patternfly-4-overrides.scss#L5-L23) which looks the same, but was split and merged into tabs & nav upstream. Meanwhile, I think we're also [overriding it in our listing panel widget, with ct-listing-panel](https://github.com/cockpit-project/cockpit/blob/main/pkg/lib/cockpit-components-listing-panel.scss#L3
), which looks the same as tabs in an expanded list.

garrett commented:
I added a patch that fixes tabs in Cockpit-Podman @ https://github.com/cockpit-project/cockpit-podman/pull/897/commits/4422e1a87c4446c8f5eb5491690ebcd056935dd3.

It's literally a one liner that can be summarized as:

```css
.pf-c-tabs__list { flex-wrap: wrap; }
```

It says "if there isn't enough space in the flex direction, then wrap if you need to. In the PF4 implementation of tabs, this is in the x direction (left-to-right) and the chonky overflow arrows don't show up as the content doesn't get truncated, so, instead, things overflow.

_Ideally_, we'd have this in PF4 overrides file and pull that in to Cockpit-Podman. (We should still have this in our overrides file as well, though. And it should apply in other places, such as Cockpit-Machines.)

Here's a stress test by using iPhone SE widths:

![Screenshot 2022-05-17 at 16-05-28 Podman containers - garrett@Bolt](https://user-images.githubusercontent.com/10246/168830426-d9dd85c4-5fec-4c00-ba5e-1bd69719adf1.png)

![Screenshot 2022-05-17 at 16-06-11 Podman containers - garrett@Bolt](https://user-images.githubusercontent.com/10246/168830438-c99786ef-c4c9-4423-87ed-84cb0e7f2e6c.png)

garrett commented:
For a direct comparison of before and after the fix, here are crops from the create modal in an iPhone SE width.

PF behavior for overflow:

![image](https://user-images.githubusercontent.com/10246/168841988-0b33cf5f-6b1f-4148-b7d1-7acde040bf7b.png)

Our wrap behavior:

![image](https://user-images.githubusercontent.com/10246/168842032-c514513b-9cf1-4bfe-97eb-e375a18897ed.png)


PF for the expanded content (and the content runs over, where some isn't even visible)

![image](https://user-images.githubusercontent.com/10246/168842906-7c1e93fa-25d9-4243-8f48-2d77f76886bf.png)

Our wrap behavior (tabs are all visible and content fits):

![image](https://user-images.githubusercontent.com/10246/168842495-f69c1451-ffe4-41f5-8810-90d1462a844b.png)

jelly commented:
PR looks fine to me just needs some fixes.

marusak commented:
Rabased, dropped commit from Garrett and replaced it with checkout of cockpit overrides.

> PR looks fine to me just needs some fixes.

Could you be more specific what fixes? :)

jelly commented on test/check-application:
```diff
@@ -1290,8 +1294,8 @@ class TestApplication(testlib.MachineCase):
         # Create Container, image is pulled and should end up being "running"
-        b.click('.pf-c-modal-box__footer #create-image-create-run-btn')
-        sel = "> span:not(.downloading)"
+        b.click('.pf-c-modal-box__footer button:contains("Create and run")')
```
What makes you prefer using `:contains`?

marusak replied:
bad rebase? :thinking:

jelly commented on test/check-application:
```diff
@@ -1853,6 +1859,82 @@ class TestApplication(testlib.MachineCase):
+        b.wait_visible('#container-details-healthcheck dt:contains("Failing streak")')
+        self.assertGreater(int(b.text('#container-details-healthcheck dt:contains("Failing streak") + dd')), 3)
+
+    @testlib.skipImage("Ubuntu-stable does not get healt check run signals", "ubuntu-stable")
```
health

jelly replied:
Is there an issue somewhere about this or is ubuntu-stable using an older podman?

marusak replied:
yeah, I did not report this as this is related to just stable version having older podman

jelly commented on test/check-application:
```diff
@@ -1853,6 +1859,82 @@ class TestApplication(testlib.MachineCase):
+    def testHealthcheckSystem(self):
+        self._testHealthcheck(True)
+
+    @testlib.skipImage("Ubuntu-stable does not get healt check run signals", "ubuntu-stable")
```
health

jelly commented on test/check-application:
```diff
@@ -1853,6 +1859,82 @@ class TestApplication(testlib.MachineCase):
+        b.set_input_text('#run-image-healthcheck-timeout input', '35')
+        b.set_input_text('#run-image-healthcheck-start-period input', '5')
+        b.click('#run-image-healthcheck-retries button:nth-child(1)')
+        b.wait_val("#run-image-healthcheck-retries input", 2)
```
should we make a new pixel test here?

jelly reviewed (approved):
Only found some small nitpicks, overall it works well!

martinpitt commented:
FTR, the fcos failure is not the fault of this PR -- it still worked yesterday in https://github.com/cockpit-project/bots/pull/3390 , and spontaneously broke. We see it in #987 and #978 as well. Must be something about these three digits..

garrett reviewed (changes requested):
Tabs are broken again...

![image](https://user-images.githubusercontent.com/10246/169784244-a1ba8a12-1a35-4fc4-bd94-36d879d88c7a.png)

I did delete and re-check out the branch again and they're still broken. (Screenshot is after that.)

marusak commented:
> I did delete and re-check out the branch again and they're still broken. (Screenshot is after that.)

Did you delete with `git clean -dffx`? (double f). You need a new `src/lib` checkout for it to work. They wrap here just fine.

garrett reviewed (approved):
Ah, I thought I did a clean during the rebuild. I guess I didn't? Anyway, after doing the ff clean, it's working.

I still think we should do a follow-up to improve the layout in smaller sizes, but that's probably more on me, once we finally make the gutter get dropped on mobile (like we do on the rest of Cockpit), as this will already solve some of the layout woes.

For this, PR: :+1: Approved!
